### PR TITLE
KAFKA-14367; Add `Heartbeat` to the new `GroupCoordinator` interface

### PR DESCRIPTION
**Pull request by dajac**
This patch adds `heartbeat` to the new `GroupCoordinator` interface and updates `KafkaApis` to use it.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 1]
```diff
@@ -2806,6 +2806,100 @@ class KafkaApisTest {
     }
   }
 
+  @ParameterizedTest
```
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:2828]: Were there previously no tests for this api? 😅

Comment (dajac) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:2828]: Yeah... At least no unit tests here.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -1764,42 +1764,38 @@ class KafkaApis(val requestChannel: RequestChannel,
...
-        val response = new HeartbeatResponse(
-          new HeartbeatResponseData()
-            .setThrottleTimeMs(requestThrottleMs)
-            .setErrorCode(error.code))
-        trace("Sending heartbeat response %s for correlation id %d to client %s."
-          .format(response, request.header.correlationId, request.header.clientId))
+    def sendResponse(response: AbstractResponse): Unit = {
+      trace("Sending heartbeat response %s for correlation id %d to client %s."
+        .format(response, request.header.correlationId, request.header.clientId))
+      requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs => {
```
Comment (jeffkbkim) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1771]: i'm assuming we'll be rebasing to use the new sendResponse right? like in https://github.com/apache/kafka/pull/12847/commits/57406651dfa3c1536e184865002e7340b7c69d55#diff-cc056b4960ededba37a438b1454f0f3c5ff5e8ad5e6d2ec9a08e813ca056ffebR1666

**Review by jeffkbkim — APPROVED** [review]
LGTM, left a small comment. `testPrepareJoinAndRejoinAfterFailedRebalance()` passed locally

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1771]: Yeah -- I left this comment on all the other PRs. David said he would rebase.

**Review by jolshan — APPROVED** [review]

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 3]
```diff
@@ -1764,42 +1764,38 @@ class KafkaApis(val requestChannel: RequestChannel,
...
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(requestThrottleMs: Int): AbstractResponse = {
-        val response = new HeartbeatResponse(
-          new HeartbeatResponseData()
-            .setThrottleTimeMs(requestThrottleMs)
-            .setErrorCode(error.code))
-        trace("Sending heartbeat response %s for correlation id %d to client %s."
-          .format(response, request.header.correlationId, request.header.clientId))
+    def sendResponse(response: AbstractResponse): Unit = {
+      trace("Sending heartbeat response %s for correlation id %d to client %s."
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Can we drop this? If trace is enabled, we should have request logging, which is more verbose.

**Review by hachikuji — APPROVED** [review]
LGTM. Left one minor comment.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 3 above)
Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Oops. We added trace logs to all the responses

Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I agree that it makes sense to remove it as we have the request log. Let me remove them everywhere.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1771]: I may address this separately as I have a +1 from Jason for this PR.

**Pull request merged**